### PR TITLE
add collection and pattern matching support to `--import.path` option

### PR DESCRIPTION
**Pull request by Toilal**
**What this PR does / why we need it**:

This replace #642 with an easier implementation.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

**Review by jkroepke — CHANGES_REQUESTED** [review]
Hi @toilal

thanks again for the new PR! I checkout it locally and it works great. But I found a new issue, that also would be happens on the PR before:

By default in Java, hidden files are included by default. As I system operator, I would not expect this behaivor. For example if you create a folder named `import-files/import-wildcard/.hidden` and place some files there, they also would include and tests will failing. If you are on a shell using globbing, hidden files are excluded by default.

Looking at the java [docs](https://docs.oracle.com/javase/8/docs/api/java/nio/file/FileSystem.html#getPathMatcher-java.lang.String-), this is an excepted behavior. 

> Leading period/dot characters in file name are treated as regular characters in match operations. For example, the "*" glob pattern matches file name ".login". The [Files.isHidden(java.nio.file.Path)](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#isHidden-java.nio.file.Path-) method may be used to test whether a file is considered hidden.

I guess a filter is needed to check, if a file or directory is is hidden. For example a folder contains a `.git` directory, keycloak-config-cli would look into it which leads into unexpected error.

File: README.md [hunk 1]
```diff
@@ -220,7 +220,7 @@ Checkout helm docs about [chart dependencies](https://helm.sh/docs/topics/charts
...
 | --keycloak.availability-check.timeout                 | KEYCLOAK_AVAILABILITYCHECK_TIMEOUT                 | Wait timeout for keycloak availability check                                      | `120s`      |                                                                                                                                 |
-| --import.path                                         | IMPORT_PATH                                        | Location of config files (if location is a directory, all files will be imported) | `/config`   | [Spring ResourceLoader](https://docs.spring.io/spring-framework/docs/current/reference/html/core.html#resources-resourceloader) |
+| --import.path                                         | IMPORT_PATH                                        | Location of config files (URL, file path, directory, or Ant-style pattern)        | `/config`   | [Spring ResourceLoader](https://docs.spring.io/spring-framework/docs/current/reference/html/core.html#resources-resourceloader) |
```
[the remark this answers is not in the document]
Comment (Toilal) [reply README.md:223]: This bring `PathMatchingResourcePatternResolver` instance into the Spring DI Context. Without this file, PathMatchingResourcePatternResolver can't be injected in other beans / services.

[the remark this answers is not in the document]
Comment (Toilal) [reply README.md:223]: Yes, I think we need a dedicated section inside readme to document behavior and possible values of `--import.path` settings.

**Comment by Toilal** [comment]
> I guess a filter is needed to check, if a file or directory is is hidden. For example a folder contains a `.git` directory, keycloak-config-cli would look into it which leads into unexpected error.

I could another --import.include-hidden-files option. I think hidden files should not be processed by default.

**Comment by jkroepke** [comment]
> I could another --import.include-hidden-files option. I think hidden files should not be processed by default.

I totally agree with you and aligned with you proposal.

**Comment by Toilal** [comment]
I'm over, could you review ? I have implemented `--import.hidden-files`` and `--import.exclude`.

**Review by jkroepke — COMMENTED** [review]
LGTM in general. Sorry for the trouble with the tests, I see a lot of force pushes here. 

- I added a suggestion around the Ant-Style pattern. No one knows what this is, except **some** java developers.
- I'm personally not happy with the order of hidden files. It feels like non-default, unexpectable. MacOS Finder, Linux Shell, IDEA sorts dot files first, then all other files. I would like to stay consistent with the 'world'. 
- Also check it locally and add some hidden files at the import-wildcard directory. Looks like tests failing again :( - Should I look into it?

And sorry again for get some points again.

File: src/main/java/de/adorsys/keycloak/config/provider/FileComparator.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+package de.adorsys.keycloak.config.provider;
+
+import java.io.File;
+import java.io.Serializable;
+import java.util.Comparator;
+
+/**
+ * Sorting comparator for files.
+ * <p>
+ * It is supported by {@link File#compareTo(File)} default behavior, but removes "." prefix of hidden files for them to be sorted naturally with
```
[the remark this answers is not in the document]
Comment (Toilal) [reply src/main/java/de/adorsys/keycloak/config/provider/FileComparator.java:30]: Yes, I don't know, maybe we should remove .dot trick, but on by ubuntu box, `ll` gives this

```
$ ll
total 48
drwxrwxr-x  2 toilal toilal 4096 févr. 24 14:22 ./
drwxrwxr-x 36 toilal toilal 4096 févr. 24 14:14 ../
-rw-rw-r--  1 toilal toilal   56 févr. 21 09:12 0_create_realm.json
-rw-rw-r--  1 toilal toilal   81 févr. 21 09:12 1_update_realm.json
-rw-rw-r--  1 toilal toilal   81 févr. 21 09:12 2_update_realm.json
-rw-rw-r--  1 toilal toilal   81 févr. 21 09:12 .3_update_realm.json
-rw-rw-r--  1 toilal toilal   81 févr. 21 09:12 4_update_realm.json
-rw-rw-r--  1 toilal toilal   81 févr. 21 09:12 5_update_realm.json
-rw-rw-r--  1 toilal toilal   81 févr. 21 09:12 6_update_realm.json
-rw-rw-r--  1 toilal toilal   81 févr. 21 09:12 .7_update_realm.json
-rw-rw-r--  1 toilal toilal   81 févr. 21 09:12 8_update_realm.json
-rw-rw-r--  1 toilal toilal   81 févr. 21 09:12 9_update_realm.json
```

Let's rollback to ordering provided by File.compareTo

File: src/main/java/de/adorsys/keycloak/config/provider/KeycloakImportProvider.java [hunk 3]
```diff
@@ -131,7 +201,7 @@ private KeycloakImport readRealmImportsFromResource(Collection<File> importResou
                         (u, v) -> {
                             throw new IllegalStateException(String.format("Duplicate key %s", u));
                         },
-                        TreeMap::new
+                        LinkedHashMap::new
```
[the remark this answers is not in the document]
Comment (Toilal) [reply src/main/java/de/adorsys/keycloak/config/provider/KeycloakImportProvider.java:204]: LinkedHashMap will keep the insertion ordering. Ordering is performed by the caller now (readPath method), so using a LinkedHashMap will keep the same ordering.

**Comment by Toilal** [comment]
> * Also check it locally and add some hidden files at the import-wildcard directory. Looks like tests failing again :( - Should I look into it?
> 
> And sorry again for get some points again.

I'll fix it, no problem. I'm using your lib since months now, maybe years, so it's time to contribute :+1:

File: src/main/java/de/adorsys/keycloak/config/provider/FileComparator.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jkroepke) [reply src/main/java/de/adorsys/keycloak/config/provider/FileComparator.java:30]: Looks like BSD ls (MacOS) and GNU ls (Linux) is different here.

**Comment by Toilal** [comment]
There's no perfect way to check that a file should be filtered out because an ancestor directory is a hidden directory.

Assume you run keycloak-config-cli in `/home/user/.kc-config` with `import.path=config-files`, it should not filter files out because o `.kc-config` being a hidden directory, as your are running keycloak-config-cli from a hidden directory !

So I think the best way is to check ancestor directories for a hidden directory backward from the resource file until PWD (current working directory) is found.

For resources outside of PWD directory, it check the whole tree path until root path.

```
PWD=/home/user/kc-config
/home/user/kc-config/a/b/c/config.json => PROCESSED
/home/.user/kc-config/a/b/c/config.json => PROCESSED (because hidden directory is an ancestor of PWD)
/home/user/kc-config/a/b/c/.config.json => NOT PROCESSED
/home/user/kc-config/a/.b/c/config.json => NOT PROCESSED
/tmp/a/.b/c/config.json => NOT PROCESSED (because hidden directory is an NOT an ancestor of PWD)
/tmp/a/b/c/config.json => PROCESSED

PWD=/home/user/.kc-config
/home/user/.kc-config/a/b/c/config.json => PROCESSED
/home/.user/.kc-config/a/b/c/config.json => PROCESSED (because hidden directory is an ancestor of PWD)
/home/user/.kc-config/a/b/c/.config.json => NOT PROCESSED
/home/user/kc-config/a/.b/c/config.json => NOT PROCESSED
/tmp/a/.b/c/config.json => NOT PROCESSED (because hidden directory is an NOT an ancestor of PWD)
/tmp/a/b/c/config.json => PROCESSED
```

**Comment by Toilal** [comment]
It Works ! :)

**Review by jkroepke — APPROVED** [review]
Thanks a lot! Looks goods!

**Pull request merged**